File: SpendingTracker-WebApp/src/api_caller/CategoryApiCaller.ts
import { Category, Option, PagedListResult } from "../types";
import { buildQueryString } from "../utils";

export async function fetchCategory(categoryId: number): Promise<Category> {
  return fetch(`/api/categories/${categoryId}`, {
    method: "GET",
    credentials: "include",
  }).then(async (response) => {
    if (!response.ok) throw new Error("Network response was not ok");
    const data = (await response.json()) as Category;
    return {
      ...data,
      createdAt: data.createdAt ? new Date(data.createdAt) : undefined,
      updatedAt: data.updatedAt ? new Date(data.updatedAt) : undefined,
    };
  });
}

export async function fetchCategoryPagedList(params: { searchString?: string; pageNumber: number; pageSize: number; }): Promise<PagedListResult<Category>> {
  const queryString = buildQueryString(params);
  return fetch(`/api/categories/get-list?${queryString}`, {
    method: "GET",
    credentials: "include",
  }).then(async (response) => {
    if (!response.ok) throw new Error("Network response was not ok");
    const data = (await response.json()) as PagedListResult<Category>;
    return {
      ...data,
      items: data.items.map((c) => ({
        ...c,
        createdAt: c.createdAt ? new Date(c.createdAt) : undefined,
        updatedAt: c.updatedAt ? new Date(c.updatedAt) : undefined,
      })),
    } as PagedListResult<Category>;
  });
}

export async function createCategory(category: Category): Promise<Category> {
  return fetch(`/api/categories/add`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    credentials: "include",
    body: JSON.stringify(category),
  }).then(async (response) => {
    if (!response.ok) throw new Error("Network response was not ok");
    const data = (await response.json()) as Category;
    return {
      ...data,
      createdAt: data.createdAt ? new Date(data.createdAt) : undefined,
      updatedAt: data.updatedAt ? new Date(data.updatedAt) : undefined,
    };
  });
}

export async function updateCategory(category: Category): Promise<Category> {
  return fetch(`/api/categories/update/${category.id}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    credentials: "include",
    body: JSON.stringify(category),
  }).then(async (response) => {
    if (!response.ok) throw new Error("Network response was not ok");
    const data = (await response.json()) as Category;
    return {
      ...data,
      createdAt: data.createdAt ? new Date(data.createdAt) : undefined,
      updatedAt: data.updatedAt ? new Date(data.updatedAt) : undefined,
    };
  });
}

export async function deleteCategory(categoryId: number): Promise<void> {
  return fetch(`/api/categories/delete/${categoryId}`, {
    method: "DELETE",
    credentials: "include",
  }).then((response) => {
    if (!response.ok) throw new Error("Network response was not ok");
  });
}

export async function fetchCategoryOptions(): Promise<Option[]> {
  return fetch("/api/categories/get-category-options", {
    method: "GET",
    credentials: "include",
  }).then(async (response) => {
    if (!response.ok) throw new Error("Network response was not ok");
    const data = (await response.json()) as Option[];
    return data;
  });
}

File: SpendingTracker-Mobile/app/src/main/java/vn/edu/fpt/spendingtracker_mobile/api_connector/CategoryApiCaller.java
package vn.edu.fpt.spendingtracker_mobile.api_connector;

import androidx.annotation.Nullable;

import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;
import retrofit2.http.Query;
import vn.edu.fpt.spendingtracker_mobile.dtos.PagedListResult;
import vn.edu.fpt.spendingtracker_mobile.entities.Category;

public interface CategoryApiCaller {
    @GET("api/categories/get/{id}")
    Call<Category> get(@Path("id") int id);

    @GET("api/categories/get-list")
    Call<PagedListResult<Category>> getPagedList(
            @Query("searchString") @Nullable String searchString,
            @Query("pageNumber") int pageNumber,
            @Query("pageSize") int pageSize
    );

    @POST("api/categories/add")
    Call<Category> add(@Body Category transactionDto);

    @PUT("api/categories/update/{id}")
    Call<Category> update(@Path("id") int id, @Body Category transactionDto);

    @DELETE("api/categories/delete/{id}")
    Call<Category> delete(@Path("id") int id);
}
